google/api_core/retry.py

```diff
@@ -301,6 +302,15 @@ class Retry(object):
         maximum (float): The maximum amount of time to delay in seconds.
         multiplier (float): The multiplier applied to the delay.
         timeout (float): How long to keep retrying, in seconds.
+        on_error (Callable[Exception]): A function to call while processing
+            a retryable exception. Any error raised by this function will
+            *not* be caught.
+        is_stream (bool): Indicates whether the input function
+            should be treated as an stream function (i.e. a Generator,
+            or function that returns an Iterable). If True, the iterable
+            will be wrapped with retry logic, and any failed outputs will
+            restart the stream. If False, only the input function call itself
```

Comment:
Restart the stream from the last non-failed output, right ?

Reply:
Since this functionality is not intended for external users directly, but rather for the library generator and handwritten library code, this seems fine.  But the generator and library developers need to be aware of this and design their `target` callable carefully if they want to accomplish certain behaviors such as the "filter through modified request" approach you described.

Please expand the docstring of `RetryableGenerator` to provide some guidance and example pattern for the library developers so they do not make false assumptions.

Reply:
Sounds good, I added a much more documentation. Let me know that looks better

So far this will mostly just be useful for hand-written layers. If we want to add this to gapic libraries automatically, we could try to classify the different kinds of streams our libraries return and provide a couple different presets for different kinds of streams. But so far that's been out of scope.